lib/broadway_kafka/producer.ex

```diff
@@ -509,7 +516,29 @@ defmodule BroadwayKafka.Producer do
       |> Keyword.put(:processors, [updated_processor_entry | other_processors_entries])
       |> Keyword.put(:batchers, updated_batchers_entries)
 
-    {allocators, updated_opts}
+    {producer_mod, producer_opts} = opts[:producer][:module]
+
+    {extra_child_specs, initialized_client_config} =
+      if producer_opts[:shared_client] do
+        client = producer_opts[:client] || BroadwayKafka.BrodClient
+
+        case client.init(Keyword.put(producer_opts, :broadway, opts)) do
+          {:error, message} ->
+            raise ArgumentError, "invalid options given to #{client}.init/1, " <> message
+
+          {:ok, config} = result ->
+            {client.shared_client_child_spec(config), result}
```

Comment:
Instead of adding one additional callback, change `client.init` to return `{:ok, child_specs, config}` or `{:error, mesage}`. :)

Reply:
The whole kafka client API is private :) we only use it for tests, feel free to break it. In any case, we can ship it as a new minor version, if we want to be safe.

Reply:
That was not clear to me at all! :rofl: . Now everything makes sense! Thanks!